haxe: add 4.1 and 4.0 compiler versions

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Fixes #95638. cc @OPNA2608 

Arguably the extlib 1.7.7 override is not ideal, but I don't really want to be stuck with it in `ocamlPackages` as long as we don't need it for anything else. The compatibility patch from 4.2 unfortunately doesn't even apply on 4.1.



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sternenseemann committed:
haxe_4_1: init at 4.1.5

sternenseemann committed:
haxe_4_0: init at 4.0.5

r-rmcgibbo commented:
Result of `nixpkgs-review pr 118112` at 108cd8e8 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>haxe_4_0</li>
    <li>haxe_4_1</li>
  </ul>
</details>

OPNA2608 commented:
Sorry for the late response! Tested `haxe_4_1` and `haxe_4_0` with a game project, both worked fine.

OPNA2608 commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -1,9 +1,23 @@
+      # haxe 4.0 and 4.1 unfortunately need an older version of extlib
+      ocaml_extlib-1-7-7 = super.ocaml_extlib.overrideAttrs (_: rec {
+        version = "1.7.7";
+        src = fetchurl {
+          url = "https://github.com/ygrek/ocaml-extlib/releases/download/${version}/extlib-${version}.tar.gz";
+          sha256 = "1sxmzc1mx3kg62j8kbk0dxkx8mkf1rn70h542cjzrziflznap0s1";
+        };
```
You could reuse `fetchFromGitHub` here.

sternenseemann replied:
No, that is a release asset and not the automatically generated archive, so we need to use `fetchurl`.

OPNA2608 replied:
Oh I see, nevermind then.

OPNA2608 reviewed (changes requested):
Tiny nitpick, otherwise looks good IMO.

sternenseemann committed:
ocamlPackages.ocaml_extlib-1-7-7: init at 1.7.7

Unfortunately there's no way to get Haxe 4.0 and 4.1 to work with extlib
1.7.8 (not even without the minimal install), so we need to package
1.7.7 again, at least until 1.7.9 (?) brings backwards compatibility
packages, hopefully.

sternenseemann commented:
@ofborg build haxe_4_0 haxe_4_1 ocamlPackages.ocaml_extlib ocamlPackages.ocaml_extlib-1-7-7
